File: day-4-face-detection/detectFaceSmileDemo.cpp
/*
Copyright 2017 BIG VISION LLC ALL RIGHTS RESERVED

This code is made available to the students of 
the online course titled "OpenCV for Beginners" 
by Satya Mallick for personal non-commercial use. 

Sharing this code is strictly prohibited without written
permission from Big Vision LLC. 

For licensing and other inquiries, please email 
spmallick@bigvisionllc.com 
*/


#include "opencv2/objdetect.hpp"
#include "opencv2/videoio.hpp"
#include "opencv2/highgui.hpp"
#include "opencv2/imgproc.hpp"

#include <iostream>
#include <stdio.h>

using namespace std;
using namespace cv;

/** Global variables */
String faceCascadePath, smileCascadePath;
CascadeClassifier faceCascade, smileCascade;

int main( int argc, const char** argv )
{
  int smileNeighborsMax = 100;
  int neighborStep = 2;

  faceCascadePath = "models/haarcascade_frontalface_default.xml";
  smileCascadePath = "models/haarcascade_smile.xml";

  //-- 1. Load the cascades
  if( !faceCascade.load( faceCascadePath ) ){ printf("--(!)Error loading face cascade\n"); return -1; };
  if( !smileCascade.load( smileCascadePath ) ){ printf("--(!)Error loading smile cascade\n"); return -1; };

  std::vector<Rect> faces;

  Mat frame = imread("hillary_clinton.jpg");
  Mat frameGray, frameClone;
  cvtColor(frame, frameGray, COLOR_BGR2GRAY);

  faceCascade.detectMultiScale( frameGray, faces, 1.4, 5);

  for ( size_t i = 0; i < faces.size(); i++ )
  {
    int x = faces[i].x;
    int y = faces[i].y;
    int w = faces[i].width;
    int h = faces[i].height;

    rectangle(frame, Point(x, y), Point(x + w, y + h), Scalar(255,0,0), 2, 4);

    Mat faceROI = frameGray( faces[i] );
    for (int neigh = 0; neigh < smileNeighborsMax; neigh = neigh + neighborStep)
    {
      std::vector<Rect> smile;

      Mat frameClone = frame.clone();
      Mat faceROIClone = frameClone( faces[i] );
      //-- In each face, detect smile
      smileCascade.detectMultiScale( faceROI, smile, 1.5, neigh);

      for ( size_t j = 0; j < smile.size(); j++ )
      {
        int smileX = smile[j].x;
        int smileY = smile[j].y;
        int smileW = smile[j].width;
        int smileH = smile[j].height;

        rectangle(faceROIClone, Point(smileX, smileY), Point(smileX + smileW, smileY + smileH), Scalar(0,255,0), 2, 4);
      }

      putText(frameClone, format("# Neighbors = %d",neigh), Point(10, 50), FONT_HERSHEY_SIMPLEX, 1, Scalar(0, 0, 255), 4);
      imshow( "Face and Smile Demo", frameClone );
      int k = waitKey(500);
      if(k ==  27)
      {
        destroyAllWindows();
        break;
      }  
    }
  }
}

File: day-4-face-detection/detectFaceSmileDemo.py
# Copyright 2017 BIG VISION LLC ALL RIGHTS RESERVED
# 
# This code is made available to the students of 
# the online course titled "OpenCV for Beginners" 
# by Satya Mallick for personal non-commercial use. 
#
# Sharing this code is strictly prohibited without written
# permission from Big Vision LLC. 
#
# For licensing and other inquiries, please email 
# spmallick@bigvisionllc.com 
#


import cv2
import sys
import numpy as np


if __name__ == '__main__':

  faceCascade = cv2.CascadeClassifier('models/haarcascade_frontalface_default.xml')
  smileCascade = cv2.CascadeClassifier('models/haarcascade_smile.xml')
  smileNeighborsMax = 100
  neighborStep = 2

  frame = cv2.imread("hillary_clinton.jpg")

  frameGray = cv2.cvtColor(frame, cv2.COLOR_BGR2GRAY)
  faces = faceCascade.detectMultiScale(frameGray, 1.4, 5)
  for (x, y, w, h) in faces:

    cv2.rectangle(frame, (x, y), (x + w, y + h), (255, 0, 0), 2)
    faceRoiGray = frameGray[y: y + h, x: x + w]
    faceRoiOriginal = frame[y: y + h, x: x + w]

    for neigh in range(1, smileNeighborsMax, neighborStep):
      smile = smileCascade.detectMultiScale(faceRoiGray, 1.5, neigh)

      frameClone = np.copy(frame)
      faceRoiClone = frameClone[y: y + h, x: x + w]
      for (xx, yy, ww, hh) in smile:
        cv2.rectangle(faceRoiClone, (xx, yy), (xx + ww, yy + hh), (0, 255, 0), 2)

      cv2.putText(frameClone, "# Neighbors = {}".format(neigh), (10, 50), cv2.FONT_HERSHEY_SIMPLEX, 1, (0, 0, 255), 4)
      cv2.imshow('Face and Smile Demo', frameClone)
      if cv2.waitKey(500) & 0xFF == 27:
        cv2.destroyAllWindows()
        sys.exit()
